### ISSUE TITLE
tf.keras.layers.Conv1D fails for RaggedTensor input

### ISSUE THREAD

**Issue by zowen** [tensorflow/tensorflow#42417]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS, probably not relevant
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary (pip)
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.7.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
I'm not sure if this is a bug or feature request, but I would expect/hope the below to work.

tf.keras.layers.Conv1D fails on a batch of type `tensorflow.python.ops.ragged.ragged_tensor.RaggedTensor`. Specifically it appears to fail on the `convert_to_tensor` step in the same manner as in https://github.com/tensorflow/tensorflow/issues/37351.

**Describe the expected behavior**

Ideally the convolution would execute as expected and produce a RaggedTensor output that reflects the shape of the input RaggedTensor and the convolution applied.

To get some insight on my use case, I am working on an unsupervised time series classification problem where the length of each input could be a giveaway and provide useless downstream embeddings. Specifically, I am using dilated convolutions to compress information about the series and using final pooling layers over the temporal dimension to produce equal size embeddings from unequal length input tensors. My first implementation used Conv1D layers with a batch size of 1, and runs fine, except that using batch size 1 limits training speed and the size of the dataset I can work with. I was hoping RaggedTensor would allow me to scale up my implementation. I suspect the lack of non-uniform sizes would preclude a GPU implementation, but I believe even using simple batching would provide a much needed speed up.

I would also be open to workarounds, if one knows of a better way to formulate this. Perhaps there is a way to operate on the padded tensor and then trim off the Conv outputs that had access to the padding as part of their input on an element-by-element basis within the batch (probably as a function of each of their true length passed in parallel)?

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow as tf
x = tf.ragged.constant([np.random.random(10+x) for x in range(5)])
x = tf.expand_dims(x, -1)
tf.keras.layers.Conv1D(7, 3)(x)
```

**Other info / logs**
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-209-09f9a8bce491> in <module>
      4 x = tf.ragged.constant([np.random.random(10+x) for x in range(5)])
      5 x = tf.expand_dims(x, -1)
----> 6 tf.keras.layers.Conv1D(7, 3)(x)

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    983 
    984         with ops.enable_auto_cast_variables(self._compute_dtype_object):
--> 985           outputs = call_fn(inputs, *args, **kwargs)
    986 
    987         if self._activity_regularizer:

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/keras/layers/convolutional.py in call(self, inputs)
    245       inputs = array_ops.pad(inputs, self._compute_causal_padding(inputs))
    246 
--> 247     outputs = self._convolution_op(inputs, self.kernel)
    248 
    249     if self.use_bias:

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    199     """Call target, and fall back on dispatchers if there is a TypeError."""
    200     try:
--> 201       return target(*args, **kwargs)
    202     except (TypeError, ValueError):
    203       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py in convolution_v2(input, filters, strides, padding, data_format, dilations, name)
   1016       data_format=data_format,
   1017       dilations=dilations,
-> 1018       name=name)
   1019 
   1020 

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py in convolution_internal(input, filters, strides, padding, data_format, dilations, name, call_from_convolution, num_spatial_dims)
   1146           data_format=data_format,
   1147           dilations=dilations,
-> 1148           name=name)
   1149     else:
   1150       if channel_index == 1:

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    199     """Call target, and fall back on dispatchers if there is a TypeError."""
    200     try:
--> 201       return target(*args, **kwargs)
    202     except (TypeError, ValueError):
    203       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    572                   func.__module__, arg_name, arg_value, 'in a future version'
    573                   if date is None else ('after %s' % date), instructions)
--> 574       return func(*args, **kwargs)
    575 
    576     doc = _add_deprecated_arg_value_notice_to_docstring(

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    572                   func.__module__, arg_name, arg_value, 'in a future version'
    573                   if date is None else ('after %s' % date), instructions)
--> 574       return func(*args, **kwargs)
    575 
    576     doc = _add_deprecated_arg_value_notice_to_docstring(

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py in conv1d(value, filters, stride, padding, use_cudnn_on_gpu, data_format, name, input, dilations)
   1887           data_format=data_format,
   1888           dilations=dilations,
-> 1889           name=name)
   1890     else:
   1891       result = squeeze_batch_dims(

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/ops/gen_nn_ops.py in conv2d(input, filter, strides, padding, use_cudnn_on_gpu, explicit_paddings, data_format, dilations, name)
    943           input, filter, strides=strides, use_cudnn_on_gpu=use_cudnn_on_gpu,
    944           padding=padding, explicit_paddings=explicit_paddings,
--> 945           data_format=data_format, dilations=dilations, name=name, ctx=_ctx)
    946     except _core._SymbolicException:
    947       pass  # Add nodes to the TensorFlow graph.

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/ops/gen_nn_ops.py in conv2d_eager_fallback(input, filter, strides, padding, use_cudnn_on_gpu, explicit_paddings, data_format, dilations, name, ctx)
   1023         "'conv2d' Op, not %r." % dilations)
   1024   dilations = [_execute.make_int(_i, "dilations") for _i in dilations]
-> 1025   _attr_T, _inputs_T = _execute.args_to_matching_eager([input, filter], ctx)
   1026   (input, filter) = _inputs_T
   1027   _inputs_flat = [input, filter]

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/eager/execute.py in args_to_matching_eager(l, ctx, default_dtype)
    265         dtype = ret[-1].dtype
    266   else:
--> 267     ret = [ops.convert_to_tensor(t, dtype, ctx=ctx) for t in l]
    268 
    269   # TODO(slebedev): consider removing this as it leaks a Keras concept.

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/eager/execute.py in <listcomp>(.0)
    265         dtype = ret[-1].dtype
    266   else:
--> 267     ret = [ops.convert_to_tensor(t, dtype, ctx=ctx) for t in l]
    268 
    269   # TODO(slebedev): consider removing this as it leaks a Keras concept.

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, dtype_hint, ctx, accepted_result_types)
   1497 
   1498     if ret is None:
-> 1499       ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
   1500 
   1501     if ret is NotImplemented:

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in _constant_tensor_conversion_function(v, dtype, name, as_ref)
    336                                          as_ref=False):
    337   _ = as_ref
--> 338   return constant(v, dtype=dtype, name=name)
    339 
    340 

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in constant(value, dtype, shape, name)
    262   """
    263   return _constant_impl(value, dtype, shape, name, verify_shape=False,
--> 264                         allow_broadcast=True)
    265 
    266 

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in _constant_impl(value, dtype, shape, name, verify_shape, allow_broadcast)
    273       with trace.Trace("tf.constant"):
    274         return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
--> 275     return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
    276 
    277   g = ops.get_default_graph()

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
    298 def _constant_eager_impl(ctx, value, dtype, shape, verify_shape):
    299   """Implementation of eager constant."""
--> 300   t = convert_to_eager_tensor(value, ctx, dtype)
    301   if shape is None:
    302     return t

~/venvs/pinpoint/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in convert_to_eager_tensor(value, ctx, dtype)
     96       dtype = dtypes.as_dtype(dtype).as_datatype_enum
     97   ctx.ensure_initialized()
---> 98   return ops.EagerTensor(value, ctx.device_name, dtype)
     99 
    100 

ValueError: TypeError: object of type 'RaggedTensor' has no len()
```

**Comment by rmothukuru** [comment]
@zowen,
Please refer this [Tensorflow Documentation for Ragged Tensor](https://www.tensorflow.org/guide/ragged_tensor#keras) which demonstrates on how to build a Model using `Ragged Tensor` and let us know if it helps. Thanks!

**Comment by zowen** [comment]
Sure, I read through that before trying anything with Ragged Tensors. Since convolutions are not conceptually bound to fixed dimensions I would hope that it could be applied to a ragged tensor, leading me to raise this bug/feature request.

**Comment by iandreariley** [comment]
@rmothukuru I agree with @zowen. It is not clear from the documentation or the exceptions whether keras convolutional ops support `RaggedTensor`. Can you clarify: Do tf.keras.Conv*D ops support `RaggedTensor` (and this is a bug) or don't they?

If they are _not_ supported I suggest the following:
1. Mention it explicitly in the docs.
2. Change the exception message to something like `TypeError: Convolution not supported for RaggedTensor`.

To add a similar use case in tf 2.3, py 3.7:
```py
import tensorflow as tf
from tensorflow.keras import models, layers

model = models.Sequential()
model.add(layers.Input([None, None, 3], dtype=tf.int64, ragged=True))
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(None, None, 3)))
```
gets
```sh
TypeError: Failed to convert object of type <class 'tensorflow.python.ops.ragged.ragged_tensor.RaggedTensor'> to Tensor. Contents: tf.RaggedTensor(values=tf.RaggedTensor(values=Tensor("Placeholder:0", shape=(None, 3), dtype=int64), row_splits=Tensor("Placeholder_2:0", shape=(None,), 
dtype=int64)), row_splits=Tensor("Placeholder_1:0", shape=(None,), dtype=int64)). Consider casting elements to a supported type.
```
Full stack trace
```
Traceback (most recent call last):
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\tensor_util.py", line 548, in make_tensor_proto
    str_values = [compat.as_bytes(x) for x in proto_values]
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\tensor_util.py", line 548, in <listcomp>       
    str_values = [compat.as_bytes(x) for x in proto_values]
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\util\compat.py", line 87, in as_bytes
    (bytes_or_text,))
TypeError: Expected binary or unicode string, got tf.RaggedTensor(values=tf.RaggedTensor(values=Tensor("Placeholder:0", shape=(None, 3), dtype=int64), row_splits=Tensor("Placeholder_2:0", shape=(None,), dtype=int64)), row_splits=Tensor("Placeholder_1:0", shape=(None,), dtype=int64)) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\Users\iriley\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\Users\iriley\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 444, in main
    run()
  File "c:\Users\iriley\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\Users\iriley\Documents\projects\Project1\simple_conv.py", line 6, in <module>
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(None, None, 3)))
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\training\tracking\base.py", line 457, in _method_wrapper 
    result = method(self, *args, **kwargs)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\keras\engine\sequential.py", line 221, in add
    output_tensor = layer(self.outputs[0])
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\keras\engine\base_layer.py", line 926, in __call__       
    input_list)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\keras\engine\base_layer.py", line 1117, in _functional_construction_call
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\keras\layers\convolutional.py", line 247, in call
    outputs = self._convolution_op(inputs, self.kernel)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\util\dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 1018, in convolution_v2
    name=name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 1148, in convolution_internal       
    name=name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 2592, in _conv2d_expanded_batch     
    name=name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\ops\gen_nn_ops.py", line 978, in conv2d
    data_format=data_format, dilations=dilations, name=name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 473, in _apply_op_helper
    raise err
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 470, in _apply_op_helper
    preferred_dtype=default_dtype)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\ops.py", line 1499, in convert_to_tensor       
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\constant_op.py", line 338, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\constant_op.py", line 264, in constant
    allow_broadcast=True)
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\constant_op.py", line 282, in _constant_impl   
    allow_broadcast=allow_broadcast))
  File "C:\Users\iriley\Anaconda3\envs\py37_tf23\lib\site-packages\tensorflow\python\framework\tensor_util.py", line 552, in make_tensor_proto
    "supported type." % (type(values), values))
TypeError: Failed to convert object of type <class 'tensorflow.python.ops.ragged.ragged_tensor.RaggedTensor'> to Tensor. Contents: tf.RaggedTensor(values=tf.RaggedTensor(values=Tensor("Placeholder:0", shape=(None, 3), dtype=int64), row_splits=Tensor("Placeholder_2:0", shape=(None,), 
dtype=int64)), row_splits=Tensor("Placeholder_1:0", shape=(None,), dtype=int64)). Consider casting elements to a supported type.
```

I can open another issue if that's necessary but it seemed better to add weight to this one.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/d0d8ca5776f525038dc3ea74fec05869/untitled63.ipynb)..Thanks !

**Comment by jvishnuvardhan** [comment]
@zowen I can reproduce the issue with `tf-nightly` (`TF2.8.dev`). 
 As it is related to keras, Is it possible for you to move this issue to keras-team/keras repo?

Please note that Keras development moved to another repository to focus entirely on only keras. Could you please repost this issue on [keras-team/keras repo](https://github.com/keras-team/keras/issues). Thanks!

**Comment by axeltidemann** [comment]
I'd also like to get an update.
